mesageEdit: Show topic input even in topic narrow.

jainkuniya opened the pull request:
Sometimes message ref is undefined, so it is not focused. Thus
use editMessage object to check for topic input visibility.

borisyankov commented:
Why do we want to show the topic input though?

PS. The complexity of the logic might need unit-tests at this point.

jainkuniya commented:
> Why do we want to show the topic input though?

Then there is no way to edit/change message topic.
Many users are facing issue of accidentally sending message to wrong topic on mobile. So this will be helpful.

> PS. The complexity of the logic might need unit-tests at this point.

Extracted logic and added tests.

borisyankov commented:
OK, cool. That makes sense.

We can definitely improve the whole approach to editing a message from a coding perspective, it is messy, but that is another PR for another time. We should do it as a separate route so going back works naturally.

gnprice commented:
Thanks @jainkuniya !

Let's keep the `getCanSelectTopic` logic in the same file/module -- it only makes sense to read it in the context of `ComposeBox`, so the extra indirection of going to another file (starting from either one) to read what's happening just gets in the way. A separate function is good because it enables tests, though.

As this Boolean logic gets more complex, another useful technique that will help keep it readable is to use `if` and early `return` to break it up -- this is another thing we gain from splitting into a function. So something like
```
if (editMessage) {
  return isStreamOrTopicNarrow(narrow);
}
if (!isStreamNarrow(narrow)) {
  return false;
}
return isMessageFocused || isTopicFocused;
```

gnprice commented:
Merged -- thanks!

For clarity, I separated out the commit that does the refactor from the one that actually makes the behavior change. Take a look.
